File: clients/js/pail-client.mjs
const CORPUS_ID = /^[a-f0-9]{32}$/;

export class PailClient {
  constructor({ baseUrl, token = "", timeoutMs = 15000, fetchImpl = globalThis.fetch } = {}) {
    if (!baseUrl) throw new TypeError("baseUrl is required");
    if (typeof fetchImpl !== "function") throw new TypeError("fetch implementation is required");
    this.baseUrl = new URL(baseUrl);
    this.token = token;
    this.timeoutMs = timeoutMs;
    this.fetch = fetchImpl;
  }

  health() {
    return this.#request("/api/health", { method: "GET" });
  }

  capabilities() {
    return this.#request("/api/capabilities", { method: "GET" });
  }

  query({ corpusId, query, limit = 5 }) {
    if (!CORPUS_ID.test(String(corpusId || ""))) throw new TypeError("corpusId must be 32 lowercase hex characters");
    const text = String(query || "").trim();
    if (!text || text.length > 4096) throw new TypeError("query must contain 1 to 4096 characters");
    if (!Number.isInteger(limit) || limit < 1 || limit > 10) throw new TypeError("limit must be an integer from 1 to 10");
    return this.#request("/api/query", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ corpus_id: corpusId, query: text, limit })
    });
  }

  async #request(path, options) {
    const headers = new Headers(options.headers || {});
    if (this.token) headers.set("authorization", `Bearer ${this.token}`);
    const response = await this.fetch(new URL(path, this.baseUrl), {
      ...options,
      headers,
      signal: AbortSignal.timeout(this.timeoutMs)
    });
    const body = await response.json().catch(() => ({ error: "invalid_json_response" }));
    if (!response.ok) {
      const error = new Error(body.error || body.reason || `PAIL request failed with ${response.status}`);
      error.status = response.status;
      error.body = body;
      throw error;
    }
    return body;
  }
}

File: clients/python/pail_client.py
"""Dependency-free client for the public PAIL gateway contract."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from typing import Any
from urllib.error import HTTPError
from urllib.request import Request, urlopen

_CORPUS_ID = re.compile(r"^[a-f0-9]{32}$")


class PailError(RuntimeError):
    def __init__(self, message: str, status: int | None = None, body: Any = None) -> None:
        super().__init__(message)
        self.status = status
        self.body = body


@dataclass(frozen=True)
class PailClient:
    base_url: str
    token: str = ""
    timeout_seconds: float = 15.0

    def health(self) -> dict[str, Any]:
        return self._request("GET", "/api/health")

    def capabilities(self) -> dict[str, Any]:
        return self._request("GET", "/api/capabilities")

    def query(self, *, corpus_id: str, query: str, limit: int = 5) -> dict[str, Any]:
        if not _CORPUS_ID.fullmatch(corpus_id):
            raise ValueError("corpus_id must be 32 lowercase hex characters")
        text = query.strip()
        if not text or len(text) > 4096:
            raise ValueError("query must contain 1 to 4096 characters")
        if not isinstance(limit, int) or not 1 <= limit <= 10:
            raise ValueError("limit must be an integer from 1 to 10")
        return self._request(
            "POST",
            "/api/query",
            {"corpus_id": corpus_id, "query": text, "limit": limit},
        )

    def _request(self, method: str, path: str, payload: Any = None) -> dict[str, Any]:
        url = self.base_url.rstrip("/") + path
        headers = {"accept": "application/json"}
        data = None
        if payload is not None:
            data = json.dumps(payload, separators=(",", ":")).encode("utf-8")
            headers["content-type"] = "application/json"
        if self.token:
            headers["authorization"] = f"Bearer {self.token}"
        request = Request(url, data=data, headers=headers, method=method)
        try:
            with urlopen(request, timeout=self.timeout_seconds) as response:
                return json.loads(response.read().decode("utf-8"))
        except HTTPError as exc:
            raw = exc.read().decode("utf-8", errors="replace")
            try:
                body = json.loads(raw)
            except json.JSONDecodeError:
                body = {"error": "invalid_json_response"}
            raise PailError(
                body.get("error") or body.get("reason") or f"PAIL request failed with {exc.code}",
                status=exc.code,
                body=body,
            ) from exc
